File: crates/python/mirascript/vm/lib/vm_global/math_unary.py
import math
from .._helpers import _expect_number, _required
from mirascript.vm.types.const import Uninitialized
import decimal


def _build(
    func, nan=None, inf=None, neginf=None, poszero=None, negzero=None, except_inf=None
):
    def wrapper(x=Uninitialized):
        x = _expect_number("x", x)
        if math.isnan(x):
            if nan is not None:
                return nan
        elif math.isinf(x):
            if x < 0 and neginf is not None:
                return neginf
            elif x > 0 and inf is not None:
                return inf
        elif x == 0.0:
            if math.copysign(1.0, x) < 0:
                if negzero is not None:
                    return negzero
            if poszero is not None:
                return poszero
        try:
            ret = func(x)
            if type(ret) is decimal.Decimal:
                ret = ret.quantize(decimal.Decimal("1"), rounding=decimal.ROUND_HALF_UP)
            return float(ret)
        except Exception as e:
            if except_inf == True:
                if x > 0:
                    return math.inf
                else:
                    return -math.inf
            return math.nan

    return wrapper


abs_ = _build(abs, math.nan, math.inf, math.inf, 0.0, 0.0)


def trunc(x=Uninitialized):
    ret = _build(math.trunc, math.nan, math.inf, -math.inf)(x)
    return math.copysign(ret, _expect_number("x", x))


floor = _build(math.floor, math.nan, math.inf, -math.inf, +0.0, -0.0)


def ceil(x=Uninitialized):
    ret = _build(math.ceil, math.nan, math.inf, -math.inf)(x)

    if ret == 0.0:
        return math.copysign(0.0, _expect_number("x", x))
    return ret


round_ = _build(lambda x: round(x, 0))


sign = _build(lambda v: 1 if v > 0 else -1 if v < 0 else 0, math.nan, 1, -1, +0.0, -0.0)

acos = _build(math.acos, math.nan, math.nan, math.nan)
acosh = _build(math.acosh, math.nan)
asin = _build(math.asin, math.nan, math.nan, math.nan)
asinh = _build(math.asinh, math.nan, math.inf, -math.inf)
atan = _build(math.atan, math.nan)


# atanh = build(math.atanh,math.nan,math.nan,math.nan)
def atanh(x=Uninitialized):
    x = _expect_number("x", x)
    ret = _build(math.atanh, math.nan)(x)
    if math.isnan(ret):
        if x == 1.0:
            return math.inf
        elif x == -1.0:
            return -math.inf
    if ret == 0.0:
        return math.copysign(0.0, x)
    return ret


cos = _build(math.cos, math.nan)
cosh = _build(math.cosh, math.nan)
sin = _build(math.sin, math.nan)
sinh = _build(math.sinh, math.nan)
tan = _build(math.tan, math.nan)
tanh = _build(math.tanh, math.nan)
exp = _build(math.exp, math.nan)
expm1 = _build(math.expm1, math.nan)
log = _build(math.log, math.nan, math.inf, math.nan, -math.inf, -math.inf)
log10 = _build(math.log10, math.nan, math.inf, math.nan, -math.inf, -math.inf)


# log1p = build(math.log1p,math.nan,math.inf,math.nan,0,-0)
def log1p(x=Uninitialized):
    x = _expect_number("x", x)
    ret = _build(math.log1p, math.nan, math.inf, math.nan)(x)
    if math.isnan(ret):
        if x == -1.0:
            return -math.inf
    if ret == 0.0:
        return math.copysign(0.0, x)
    return ret


log2 = _build(math.log2, math.nan, math.inf, math.nan, -math.inf, -math.inf)
sqrt = _build(math.sqrt, math.nan)
cbrt = _build(lambda x: math.copysign(abs(x) ** (1 / 3), x), math.nan)

File: packages/mirascript/src/vm/lib/global/math-unary.ts
import type { VmAny } from '../../index.js';
import { $ToNumber } from '../../operations.js';
import { required, VmLib } from '../_helpers.js';

/** 生成函数 */
function build(f: (x: number) => number): (x: VmAny) => number {
    return (x) => {
        required('x', x, Number.NaN);
        return f($ToNumber(x));
    };
}

export const trunc = VmLib(build(Math.trunc), {
    summary: '返回数值的整数部分（去除小数）',
    params: { x: '要取整数部分的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const floor = VmLib(build(Math.floor), {
    summary: '返回小于等于给定数的最大整数',
    params: { x: '要向下取整的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const ceil = VmLib(build(Math.ceil), {
    summary: '返回大于等于给定数的最小整数',
    params: { x: '要向上取整的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});

const _round = Math.round;
const _abs = Math.abs;
export const round = VmLib(
    build((x) => {
        // Ref: https://github.com/python/cpython/blob/9ce99c6c1901705238e4cb3ce81eb6f499e7b4f4/Objects/floatobject.c#L1052-L1060
        let rounded = _round(x);
        if (_abs(x - rounded) === 0.5) {
            // 使用银行家舍入法
            rounded = 2 * _round(x / 2);
        }
        return rounded;
    }),
    {
        summary: '返回四舍五入后的整数',
        params: { x: '要四舍五入的数' },
        paramsType: { x: 'number' },
        returnsType: 'number',
    },
);
export const sign = VmLib(build(Math.sign), {
    summary: '返回数值的符号（正数为 1，负数为 -1，零为 0）',
    params: { x: '要判断符号的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const abs = VmLib(build(Math.abs), {
    summary: '返回数值的绝对值',
    params: { x: '要取绝对值的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});

export const acos = VmLib(build(Math.acos), {
    summary: '返回数值的反余弦值（弧度）',
    params: { x: '要计算反余弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const acosh = VmLib(build(Math.acosh), {
    summary: '返回数值的反双曲余弦值',
    params: { x: '要计算反双曲余弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const asin = VmLib(build(Math.asin), {
    summary: '返回数值的反正弦值（弧度）',
    params: { x: '要计算反正弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const asinh = VmLib(build(Math.asinh), {
    summary: '返回数值的反双曲正弦值',
    params: { x: '要计算反双曲正弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const atan = VmLib(build(Math.atan), {
    summary: '返回数值的反正切值（弧度）',
    params: { x: '要计算反正切的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const atanh = VmLib(build(Math.atanh), {
    summary: '返回数值的反双曲正切值',
    params: { x: '要计算反双曲正切的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const cos = VmLib(build(Math.cos), {
    summary: '返回数值的余弦值',
    params: { x: '要计算余弦的数（弧度）' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const cosh = VmLib(build(Math.cosh), {
    summary: '返回数值的双曲余弦值',
    params: { x: '要计算双曲余弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const sin = VmLib(build(Math.sin), {
    summary: '返回数值的正弦值',
    params: { x: '要计算正弦的数（弧度）' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const sinh = VmLib(build(Math.sinh), {
    summary: '返回数值的双曲正弦值',
    params: { x: '要计算双曲正弦的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const tan = VmLib(build(Math.tan), {
    summary: '返回数值的正切值',
    params: { x: '要计算正切的数（弧度）' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const tanh = VmLib(build(Math.tanh), {
    summary: '返回数值的双曲正切值',
    params: { x: '要计算双曲正切的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});

export const exp = VmLib(build(Math.exp), {
    summary: '返回 e 的指定次幂',
    params: { x: '指数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const expm1 = VmLib(build(Math.expm1), {
    summary: '返回 e 的 x 次幂减 1',
    params: { x: '指数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const log = VmLib(build(Math.log), {
    summary: '返回数值的自然对数（以 e 为底）',
    params: { x: '要取对数的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const log10 = VmLib(build(Math.log10), {
    summary: '返回数值的以 10 为底的对数',
    params: { x: '要取对数的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const log1p = VmLib(build(Math.log1p), {
    summary: '返回 1 加上数值的自然对数',
    params: { x: '要取对数的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const log2 = VmLib(build(Math.log2), {
    summary: '返回数值的以 2 为底的对数',
    params: { x: '要取对数的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});

export const sqrt = VmLib(build(Math.sqrt), {
    summary: '返回数值的平方根',
    params: { x: '要开平方的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
export const cbrt = VmLib(build(Math.cbrt), {
    summary: '返回数值的立方根',
    params: { x: '要计算立方根的数' },
    paramsType: { x: 'number' },
    returnsType: 'number',
});
